lib/Dialect/TTIR/IR/TTIROps.cpp

```diff
@@ -409,34 +414,149 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
   ::mlir::RankedTensorType inputAType = getA().getType();
   ::mlir::RankedTensorType inputBType = getB().getType();
   ::mlir::RankedTensorType outputType = getOutput().getType();
-  auto inputAShape = inputAType.getShape();
-  auto inputBShape = inputBType.getShape();
-  auto outputShape = outputType.getShape();
-  if (inputAShape.size() < 2) {
-    return emitOpError("Input A must be at least a 2D tensor");
+
+  llvm::ArrayRef<int64_t> outputShape = outputType.getShape();
+  llvm::SmallVector<int64_t> inputAShape, inputBShape;
+
+  // Assemble the shape of the input tensor A
+  for (int i = 0; i < inputAType.getRank(); i++) {
+    inputAShape.push_back(inputAType.getShape()[i]);
   }
-  if (inputBShape.size() < 2) {
-    return emitOpError("Input B must be at least a 2D tensor");
+
+  // If input A is a vector (1D tensor), 1 is prepended to its dimension for the
+  // purpose of the matrix multiply. After the matrix multiply, the prepended
+  // dimension is removed.
+  if (inputAType.getRank() == 1) {
+    inputAShape.insert(inputAShape.begin(), 1);
+  }
+
+  // Assemble the shape of the input tensor B
+  for (int i = 0; i < inputBType.getRank(); i++) {
+    inputBShape.push_back(inputBType.getShape()[i]);
+  }
+
+  // If input B is a vector (1D tensor), a 1 is appended to its dimension for
+  // the purpose of the matrix-vector product and removed after.
+  if (inputBType.getRank() == 1) {
+    inputBShape.push_back(1);
   }
-  if (inputAShape.size() != inputBShape.size()) {
-    return emitOpError("Input A and B must have the same rank");
+
+  // Verify that the input A is at least 1D tensor
+  if (inputAType.getRank() < 1) {
+    return emitOpError("Input A must be at least a 1D tensor");
   }
-  if (inputAShape.size() != outputShape.size()) {
-    return emitOpError("Input A and B must have the same rank as the output");
+
+  // Verify that the input B is at least 1D tensor
+  if (inputBType.getRank() < 1) {
+    return emitOpError("Input B must be at least a 1D tensor");
   }
+
+  // Verify that the input A and input B has matching inner dimensions
   if (inputAShape[inputAShape.size() - 1] !=
       inputBShape[inputBShape.size() - 2]) {
-    return emitOpError("Input A and B must have matching inner dimensions");
+    return emitOpError(
+        "Input A[-1](" + std::to_string(inputAShape[inputAShape.size() - 1]) +
+        ") and B[-2](" + std::to_string(inputBShape[inputBShape.size() - 2]) +
+        ") must have matching inner dimensions");
   }
-  if (outputShape[outputShape.size() - 2] !=
-      inputAShape[inputAShape.size() - 2]) {
-    return emitOpError("Output must have the same number of rows as input A");
```

Comment:
Sometimes for these kinds of checks, it can be simpler to normalize the rank of both inputs first:
```c++
while (inputAShape.size() < inputBShape.size) inputAShape.insert(..., 1);
while (inputBShape.size() < inputAShape.size) inputBShape.insert(..., 1);
```

Now you can compare respective dims in lockstep with each other and assert for equality or if one or the other is `1` for legal bcastable dims.  It's then trivial to form an output shape.

Reply:
At this particular place, I am checking the inner dimensions. After that, I created the output expectedOutputShape which later I checked against the provided output shape.

I tried to reuse the llvm's OpTrait::util::getBroadcastedShape to get the broadcast check and form the output expectedOutputShape. Maybe I can rewrite the code for checking the broadcast more simpler without it, but not sure.

Reply:
I created a join function in utils.h, it helps a lot to shorten the code and makes it much more readable now.